### PR TITLE
A typed search engine

### PR DESCRIPTION
**Pull request by gamburgm**
Add types to our search engine.

Future work: 
1. add types to our DB + serializers—will happen with new ORM
2. unify the frontend and backend on shared types

### CODE REVIEW

File: backend/types.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  }
+};
+
+export interface EsResultBody {
+  took: number,
+  hits: {
+    total: { value: number },
+    hits: {
+      // TODO blah
+      // EsSerializer types
```
Comment (dajinchu) [inline backend/types.ts]: can you steal the types from `scrapers-cleanup` that is now in master?

Comment (gamburgm) [reply backend/types.ts]: Yeah that's the plan

File: backend/search_types.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  [fieldName: string]: EsValue;
+}
+// ====== Bool Queries =======
+export interface BoolQuery {
+  bool: BoolType;
+}
+
+export type BoolType = MustQuery | ShouldQuery | FilterQuery;
+
+// THESE ARE THE SAME. The only difference is what string shows up first.
```
Comment (dajinchu) [inline backend/search_types.ts]: confusing comment

File: backend/search_types.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  class: Course;
+  sections: Section[];
+}
+export type ResultDoc = Employee | CourseDoc;
+
+/**************** ES UTILS *****************/
+export type EsMapping = Record<string, any>;
+
+// TODO value is a Document--something we insert into ES
+export type EsBulkData = Record<string, any>;
```
Comment (dajinchu) [inline backend/search_types.ts:222]: is ResultDoc not the same as the docs in bulkdata?

Comment (dajinchu) [reply backend/search_types.ts:222]: oh nvm i see, ResultDoc is from postgres

**Review by dajinchu — APPROVED** [review]

**Pull request merged**